compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -132,15 +132,23 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
     let mut file = write::Object::new(binary_format, architecture, endianness);
     match architecture {
         Architecture::Mips => {
-            // copied from `mipsel-linux-gnu-gcc foo.c -c` and
-            // inspecting the resulting `e_flags` field.
-            let e_flags = elf::EF_MIPS_CPIC
-                | elf::EF_MIPS_PIC
-                | if sess.target.options.cpu.contains("r6") {
-                    elf::EF_MIPS_ARCH_32R6 | elf::EF_MIPS_NAN2008
-                } else {
-                    elf::EF_MIPS_ARCH_32R2
-                };
+            let arch = match sess.target.options.cpu.as_ref() {
+                "mips1" => elf::EF_MIPS_ARCH_1,
+                "mips2" => elf::EF_MIPS_ARCH_2,
+                "mips3" => elf::EF_MIPS_ARCH_3,
+                "mips4" => elf::EF_MIPS_ARCH_4,
+                "mips5" => elf::EF_MIPS_ARCH_5,
+                s if s.contains("r6") => elf::EF_MIPS_ARCH_32R6,
+                _ => elf::EF_MIPS_ARCH_32R2,
+            };
+            // The only ABI LLVM supports for 32-bit MIPS CPUs is o32.
+            let mut e_flags = elf::EF_MIPS_CPIC | elf::EF_MIPS_ABI_O32 | arch;
+            if sess.target.options.relocation_model != RelocModel::Static {
+                e_flags |= elf::EF_MIPS_PIC;
```

Comment:
`fn create_object_file` is only used for creating object files that don't contain any actual code, only metadata or undefined symbol references.
I wonder which of these flags are really necessary.

It's clear that the ABI bits need to match between all the linked object files.
But what happens if you remove the PIC bits, for example? (Or arch, or nan, etc.)

Reply:
I agree it shouldn't matter, but lld doesn't check if object files have no code before checking for mismatched flags (I assume for efficiency).

Mismatched archs [fail here](https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/MipsArchTree.cpp#L277-L292), PIC [fails here](https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/MipsArchTree.cpp#L104-L114) and [NaN here](https://github.com/llvm/llvm-project/blob/main/lld/ELF/Arch/MipsArchTree.cpp#L81-L84). FWIW the nan flag was [added after](https://github.com/rust-lang/rust/commit/579e8bce7df948c786ab89f2a5d4b949cf44a44d) the initial PR for generating this object file, so I assume this has been an issue on targets other than the two I mentioned.